### ISSUE TITLE
Major performance issue using `AsSplitQuery` in EF Core 5 after upgrading from EF Core 2

### ISSUE THREAD

**Issue by Obi-Dann** [dotnet/efcore#24808]
Hello,

We've been attempting to upgrade our EF Core 2 to EF Core 5 and had to revert because there was a major performance degradation caused by changes in how split queries are generated.

We have queries with lots of nested `Include()`. Some of those queries used to return 1K+ rows with EF Core 2, but the row count exploded to 100K+ rows with EF Core 5 basically killing the app.

There's a repo with replication https://github.com/Obi-Dann/EfCoreSplitQueryPerformanceIssue

The repo contains a simple DB Model showcasing the performance issue
<details>
<summary>Database diagram</summary>

![image](https://user-images.githubusercontent.com/12251341/116647619-14f72e00-a9cf-11eb-93f4-b1e95538329e.png)

</details>

When I query `Events` including `Participants`, `CreatedBy` and `Hobbies`, EF Core 2 makes 3 queries which return
1 `Event`, 3 `Participants` and 3 `Hobbies`.

<details>
<summary>EF Core 2 output</summary>

```
EF Core Version 2.2.6-servicing-10079
Executed DbCommand (0ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
SELECT "x"."EventId", "x"."EventName"
FROM "Events" AS "x"
WHERE "x"."EventId" = 1
ORDER BY "x"."EventId"

Results:
----------------------- 
| EventId | EventName |
----------------------- 
| 1       | Event 1   |
----------------------- 

Count: 1

Executed DbCommand (0ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
SELECT "x.Participants"."ParticipantId", "x.Participants"."CreatedById", "x.Participants"."EventId", "x.Participants"."ParticipantName", "p.CreatedBy"."UserId", "p.CreatedBy"."UserName"
FROM "Participants" AS "x.Participants"
INNER JOIN "Users" AS "p.CreatedBy" ON "x.Participants"."CreatedById" = "p.CreatedBy"."UserId"
INNER JOIN (
SELECT "x0"."EventId"
FROM "Events" AS "x0"
WHERE "x0"."EventId" = 1
) AS "t" ON "x.Participants"."EventId" = "t"."EventId"
ORDER BY "t"."EventId", "p.CreatedBy"."UserId"

Results:
------------------------------------------------------------------------------- 
| ParticipantId | CreatedById | EventId | ParticipantName | UserId | UserName |
------------------------------------------------------------------------------- 
| 1             | 1           | 1       | Participant 1   | 1      | User 1   |
------------------------------------------------------------------------------- 
| 2             | 1           | 1       | Participant 2   | 1      | User 1   |
------------------------------------------------------------------------------- 
| 3             | 1           | 1       | Participant 3   | 1      | User 1   |
------------------------------------------------------------------------------- 

Count: 3

Executed DbCommand (0ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
SELECT "p.CreatedBy.Hobbies"."HobbyId", "p.CreatedBy.Hobbies"."HobbyName", "p.CreatedBy.Hobbies"."UserId"
FROM "Hobbies" AS "p.CreatedBy.Hobbies"
INNER JOIN (
SELECT DISTINCT "p.CreatedBy0"."UserId", "t0"."EventId"
FROM "Participants" AS "x.Participants0"
INNER JOIN "Users" AS "p.CreatedBy0" ON "x.Participants0"."CreatedById" = "p.CreatedBy0"."UserId"
INNER JOIN (
SELECT "x1"."EventId"
FROM "Events" AS "x1"
WHERE "x1"."EventId" = 1
) AS "t0" ON "x.Participants0"."EventId" = "t0"."EventId"
) AS "t1" ON "p.CreatedBy.Hobbies"."UserId" = "t1"."UserId"
ORDER BY "t1"."EventId", "t1"."UserId"

Results:
--------------------------------- 
| HobbyId | HobbyName  | UserId |
--------------------------------- 
| 1       | Crocheting | 1      |
--------------------------------- 
| 2       | Beatboxing | 1      |
--------------------------------- 
| 3       | Witchcraft | 1      |
--------------------------------- 

Count: 3
```

</details>

However, EF Core 5 and 6 return much larger result set:
1 `Event`, 3 `Participants` and **9** `Hobbies`.

<details>
<summary>EF Core 5 output</summary>

```
EF Core Version 5.0.5
Executed DbCommand (0ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
SELECT "e"."EventId", "e"."EventName"
FROM "Events" AS "e"
WHERE "e"."EventId" = 1
ORDER BY "e"."EventId"

Results:
 ----------------------- 
 | EventId | EventName |
 ----------------------- 
 | 1       | Event 1   |
 ----------------------- 

 Count: 1

Executed DbCommand (0ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
SELECT "t"."ParticipantId", "t"."CreatedById", "t"."EventId", "t"."ParticipantName", "t"."UserId", "t"."UserName", "e"."EventId"
FROM "Events" AS "e"
INNER JOIN (
    SELECT "p"."ParticipantId", "p"."CreatedById", "p"."EventId", "p"."ParticipantName", "u"."UserId", "u"."UserName"
    FROM "Participants" AS "p"
    INNER JOIN "Users" AS "u" ON "p"."CreatedById" = "u"."UserId"
) AS "t" ON "e"."EventId" = "t"."EventId"
WHERE "e"."EventId" = 1
ORDER BY "e"."EventId", "t"."ParticipantId", "t"."UserId"

Results:
 ----------------------------------------------------------------------------------------- 
 | ParticipantId | CreatedById | EventId | ParticipantName | UserId | UserName | EventId |
 ----------------------------------------------------------------------------------------- 
 | 1             | 1           | 1       | Participant 1   | 1      | User 1   | 1       |
 ----------------------------------------------------------------------------------------- 
 | 2             | 1           | 1       | Participant 2   | 1      | User 1   | 1       |
 ----------------------------------------------------------------------------------------- 
 | 3             | 1           | 1       | Participant 3   | 1      | User 1   | 1       |
 ----------------------------------------------------------------------------------------- 

 Count: 3

Executed DbCommand (0ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
SELECT "h"."HobbyId", "h"."HobbyName", "h"."UserId", "e"."EventId", "t"."ParticipantId", "t"."UserId"
FROM "Events" AS "e"
INNER JOIN (
    SELECT "p"."ParticipantId", "p"."EventId", "u"."UserId"
    FROM "Participants" AS "p"
    INNER JOIN "Users" AS "u" ON "p"."CreatedById" = "u"."UserId"
) AS "t" ON "e"."EventId" = "t"."EventId"
INNER JOIN "Hobbies" AS "h" ON "t"."UserId" = "h"."UserId"
WHERE "e"."EventId" = 1
ORDER BY "e"."EventId", "t"."ParticipantId", "t"."UserId"

Results:
 -------------------------------------------------------------------- 
 | HobbyId | HobbyName  | UserId | EventId | ParticipantId | UserId |
 -------------------------------------------------------------------- 
 | 1       | Crocheting | 1      | 1       | 1             | 1      |
 -------------------------------------------------------------------- 
 | 2       | Beatboxing | 1      | 1       | 1             | 1      |
 -------------------------------------------------------------------- 
 | 3       | Witchcraft | 1      | 1       | 1             | 1      |
 -------------------------------------------------------------------- 
 | 1       | Crocheting | 1      | 1       | 2             | 1      |
 -------------------------------------------------------------------- 
 | 2       | Beatboxing | 1      | 1       | 2             | 1      |
 -------------------------------------------------------------------- 
 | 3       | Witchcraft | 1      | 1       | 2             | 1      |
 -------------------------------------------------------------------- 
 | 1       | Crocheting | 1      | 1       | 3             | 1      |
 -------------------------------------------------------------------- 
 | 2       | Beatboxing | 1      | 1       | 3             | 1      |
 -------------------------------------------------------------------- 
 | 3       | Witchcraft | 1      | 1       | 3             | 1      |
 -------------------------------------------------------------------- 

 Count: 9
```

</details>

While `6` additional rows is not a big problem, it becomes a major issue on larger datasets, essentially exploding the
number of rows returned from 1k to 100k.
It looks like every `Hobby` row has Ids of the whole graph (`EventId | ParticipantId | UserId`), but perhaps it can only contain `UserId` and use `Distinct` or something like that to stop getting duplicates?

Is there any way to avoid this performance degradation without rewriting queries? What would be an advice on rewriting the queries given the query is quite simple and should perform OK?

The repo uses Sqlite but this issue can be replicated with SQL Server.

EF Core version: 2.2.6, 5.0.5, 6.0.0-preview.5.21228.4
Database provider: Microsoft.EntityFrameworkCore.Sqlite, Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 5.0 
Operating system: macOS, Windows

**Comment by ajcvickers** [comment]
@maumar to take a look.

**Comment by maumar** [comment]
This is a legitimate perf regression for some scenarios - we redesigned a way we perform split query to support more scenarios. Previously we reverse order of the navigation chain (1st query - select customers, 2nd query - select orders join customer). In the new way we query for the navigations in the same order (1st query - select customers, 2nd query - select customers join orders) - this allows us to support more scenarios (e.g. Take). However now we can't rely on identity resolution to return fewer results.

@Obi-Dann Currently there is no way to revert to the old behavior, so the queries would need to be rewritten.

The "savings" in the previous version were coming from many to one relationships. So the queries can be rewritten at these points:

```cs
                var q1 = db.Events
                    .Include(x => x.Participants)
                    .ThenInclude(x => x.CreatedBy).ToList();

                var q2 = (from h in db.Hobbies
                          join u in db.Users on EF.Property<long>(h, "UserId") equals u.UserId
                          join p in db.Participants on u.UserId equals p.CreatedById
                          where p.EventId == 1
                          select h).ToList();
```

that second query could be easier to write if the entities have navigations pointing back:

```cs
                var q2 = db.Hobbies.Where(h => h.User.Participants.Any(p => p.EventId == 1)).ToList();
```

Basically fetch relevant event(s), participants and user(s) in the first query, and the events related with that user in the second query. Identity resolution will then stitch the navigations together.

**Comment by maulik-modi** [comment]
@Obi-Dann , Did it fix your issue?

**Comment by ajcvickers** [comment]
Note from triage: we discussed this and, while the performance of these queries has regressed, we still believe that the design change to support more types of queries is the correct way to go. This may mean some queries need to be re-written, like in the example above, to get back better performance.
